src/mainloop.cpp

```diff
@@ -154,13 +154,16 @@ int Mainloop::write_msg(const std::shared_ptr<Endpoint> &e, const struct buffer
     return r;
 }
 
-void Mainloop::route_msg(struct buffer *buf)
+void Mainloop::route_msg(const Endpoint *msg_receiver, struct buffer *buf)
 {
     bool unknown = true;
 
     for (const auto &e : this->g_endpoints) {
-        auto acceptState = e->accept_msg(buf);
+        if (e.get() == msg_receiver) {
+            continue; // don't offer message to the endpoint where it was received
+        }
```

Comment:
what case this covers what is not covered by `has_sys_comp_id()` inside Endpoint::accept_msg() that is called just below?

Reply:
It originates from the parallel data links issue and actually should replace the `has_sys_comp_id()` check. But at a second thought, it's more of a workaround now than a real fix to that issue. (I don't think discussing the details in this PR is the right place.)
I don't know, why that isn't in the commit. But then, it'll be a bigger discussion topic and I'll open another PR for that.